src/main/java/duke/application/Duke.java

```diff
@@ -0,0 +1,243 @@
+                    } catch (IllegalTodoException e) {
+                        System.out.println("OOPS!!! The description of todo cannot be empty");
+                    }
+                } else if (userInput.startsWith("delete")) {
+                    try {
+                        int indexToDelete = getIndexToDelete(userInput);
+                        Task task = taskList.get(indexToDelete);
+                        taskList.remove(indexToDelete);
+                        System.out.println("Noted. Olivia has removed this task:");
+                        System.out.println(task);
+                        printTotalNumberOfTasks(taskList);
+                    } catch (IllegalDeleteException e) {
+                        System.out.println("To delete a task, please enter in the following format: delete <Task Number>");
+                    } catch (NumberFormatException e) {
+                        System.out.println("You must enter a task number to delete it");
+                    } catch (IndexOutOfBoundsException e) {
+                        System.out.println("The task number you entered does not exist");
+                    }
+                }
+                userInput = sc.nextLine();
+                checkUserInputValidity(userInput);
+            } catch (DukeException e) {
+                printIllegalCommandErrorMessage();
+            }
+        }
+        TaskDatabase.getInstance().save("taskdata.txt",taskList);
+        printGoodbyeMessage();
+    }
+
+    private static int getIndexToDelete(String userInput) throws IllegalDeleteException {
+        String[] tokenArray = stringToToken(userInput," ");
+        if (tokenArray.length != 2) {
+            throw new IllegalDeleteException();
+        }
+        return Integer.parseInt(tokenArray[1].trim()) - 1;
+    }
+
+
+    private static void printIllegalCommandErrorMessage() {
+        System.out.println("OOPS you have confused Olivia, please enter a valid command");
+        System.out.println("Please make sure your command starts with todo, event, deadline, mark,"
+                + " unmark or equals list or bye");
+    }
+
+    private static void checkUserInputValidity(String userInput) throws DukeException {
+        //checking validity of commands which uses "start with"
+        List<String> validCommandsForStartsWith = Arrays.asList("todo", "event","deadline","mark","unmark","delete");
+        boolean isStartWithCommandValid = validCommandsForStartsWith.stream().anyMatch(userInput::startsWith);
+        //checking validity of commands which uses "equals"
+        List<String> validCommandsForEquals = Arrays.asList("list","bye");
+        boolean isEqualsToCommandValid = validCommandsForEquals.stream().anyMatch(userInput::equals);
+        //if either 1 is true -> don't throw exception, else throw an exception
+        if (!(isStartWithCommandValid ^ isEqualsToCommandValid)) {
+            throw new DukeException();
+        }
```

Comment:
This may be too complicated for people to understand, perhaps you can simplify it?